--- a/trainer.py
+++ b/trainer.py
@@ -714,6 +714,8 @@
         if args.torch_compile and not is_torch_compile_available():
             raise RuntimeError("Using torch.compile requires PyTorch 2.0 or higher.")
 
+        self.main_loss_logged = False
+
         self.is_fsdp_xla_v2_enabled = args.fsdp_config.get("xla_fsdp_v2", False)
         if self.is_fsdp_xla_v2_enabled:
             if not IS_XLA_FSDPV2_POST_2_2:
@@ -2909,6 +2911,7 @@
             self.store_flos()
 
             self.log(logs)
+            self.main_loss_logged = True
 
         metrics = None
         if self.control.should_evaluate:
